ls1intum/Artemis
# `Programming exercises`: Add grading based on programming tasks

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context


Configure grading for automatic tests individually for each test is bothersome and error prone.

By configuring the tests based on tasks instructors should be able to set the correct settings easier.

### Description

Changes the table of test cases to a table of tasks with collapsible tests.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise (default settings should suffice)

1. Go to the exercise grading tab. e.g. course management -> exercises -> grading
2. Try setting various settings for tasks and tests. See if the table updates correctly i.e. task updates the test cases it contains; test cases update the task it belongs to
3. Try the save and check if the test cases update correctly e.g. via the chart
4. Try the reset button and check if test cases are updated correctly

#### Additional edge cases:
**Test inactive test cases**
1. Create inactive test cases by commenting them out in the test repository. e.g. go to course-management -> exercises -> edit in editor -> select test repository and comment out tests
2. Go to the grading tab and see if the visibility correctly corresponds to the `show inactive test cases` toggle
3. Try the toggle and see if the table updates correctly

**Test case assigned to multiple tasks**
1. In the problem statement assign a test case to multiple tasks. This can be done by editing the markdown and adding the test case name to a task.
2. Go to the grading tab and change something in the test assigned to multiple tasks.
3. Check if all tasks update correctly and the information in the table still makes sense

### Review Progress






#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage




| Class/File | Branch | Line |
|------------|-------:|-----:|
| programming-exercise-grading-task.component.ts | 75% | 96.15% |
| programming-exercise-task.service.ts | 76.05% | 93.49% |
| programming-exercise-task.component.ts | 35% | 96.42% |
| programming-exercise-configure-grading.component.ts | 69.78% | 83.06% |

### Screenshots



![artemis-test1 artemis in tum de_course-management_66_programming-exercises_2326_grading_test-cases](https://user-images.githubusercontent.com/25489463/228879110-2fc848f8-6d85-4097-b2aa-3384d5ea587d.png)
![artemis-test1 artemis in tum de_course-management_66_programming-exercises_2326_grading_test-cases (1)](https://user-images.githubusercontent.com/25489463/228879122-611e15aa-7d4f-4141-93fb-4fca21d7daee.png)
